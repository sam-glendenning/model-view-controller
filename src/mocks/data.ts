import type { Post, User } from '@/shared/types';

// Mock data
export const mockUsers: User[] = [
  {
    id: 1,
    name: 'John Doe',
    email: 'john.doe@example.com',
    phone: '555-0123',
    website: 'johndoe.com',
    company: {
      name: 'Doe Enterprises',
      catchPhrase: 'Innovation at its best',
      bs: 'synergistic solutions',
    },
    address: {
      street: '123 Main St',
      suite: 'Apt 1',
      city: 'Anytown',
      zipcode: '12345',
      geo: {
        lat: '40.7128',
        lng: '-74.0060',
      },
    },
  },
  {
    id: 2,
    name: 'Jane Smith',
    email: 'jane.smith@example.com',
    phone: '555-0456',
    website: 'janesmith.com',
    company: {
      name: 'Smith & Co',
      catchPhrase: 'Quality first',
      bs: 'scalable platforms',
    },
    address: {
      street: '456 Oak Ave',
      suite: 'Suite 200',
      city: 'Other City',
      zipcode: '67890',
      geo: {
        lat: '34.0522',
        lng: '-118.2437',
      },
    },
  },
];

export const mockPosts: Post[] = [
  {
    id: '1',
    userId: 1,
    title: 'Test Post Title',
    body: 'This is a test post body with some content to display.',
  },
  {
    id: '2',
    userId: 1,
    title: 'Another Test Post',
    body: 'This is another test post with different content.',
  },
  {
    id: '3',
    userId: 2,
    title: "Jane's Post",
    body: 'A post written by Jane Smith about various topics.',
  },
  {
    id: '4',
    userId: 1,
    title: 'Understanding React Hooks',
    body: 'React Hooks have revolutionized how we write components. This post explores the most commonly used hooks and their practical applications.',
  },
  {
    id: '5',
    userId: 2,
    title: 'Building Scalable APIs',
    body: 'A comprehensive guide to building APIs that can handle growth and maintain performance as your application scales.',
  },
];
