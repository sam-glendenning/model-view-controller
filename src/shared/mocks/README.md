# MSW Mock Setup Architecture

This project uses different MSW handler strategies for development vs testing to address different requirements.

## Handler Types

### 1. Runtime Handlers (`handlers.ts`)

**Purpose**: Development experience with persistent data

- **Stateful**: Mutations persist during development session
- **Realistic**: Full CRUD operations modify shared state
- **Session continuity**: Data changes visible across page refreshes

**Usage**:

- Development server (`yarn dev`)
- Manual testing and UI development

### 2. Test Handlers (`testHandlers.ts`)

**Purpose**: Isolated, deterministic testing

- **Stateless**: No mutations to global mock data
- **Isolated**: Each test gets fresh data copies
- **Predictable**: Tests don't affect each other

**Usage**:

- Automated tests (`yarn test`)
- CI/CD pipelines

## Why This Separation?

### The Problem

Originally, both development and tests used the same handlers, which caused:

- **Test pollution**: One test's mutations affected subsequent tests
- **Flaky tests**: Tests passed/failed depending on execution order
- **Hard to debug**: Mysterious state changes between tests

### The Solution

- **Development**: Keep stateful handlers for realistic UX
- **Testing**: Use stateless handlers for test isolation
- **Clean separation**: Different needs, different implementations

## Best Practices

### For Real Applications

In production apps:

- **Development**: Use MSW for consistent development experience
- **Testing**: Use MSW with isolated handlers (like this project)
- **Production**: Real API calls (no MSW)

### Handler Guidelines

- **Runtime handlers**: Mutate shared state for persistence
- **Test handlers**: Return isolated copies, never mutate globals
- **Always**: Deep clone mock data in test handlers
