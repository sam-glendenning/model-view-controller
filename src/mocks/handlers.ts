import { http, HttpResponse, delay } from 'msw';
import type { Post } from '@/shared/types';
import { mockUsers, mockPosts } from './data';
import { v4 as uuidv4 } from 'uuid';

// Artificial delay function
const artificialDelay = () => delay(1000);

// Request handlers with artificial delays
export const handlers = [
  // Users
  http.get('https://jsonplaceholder.typicode.com/users', async () => {
    await artificialDelay();
    return HttpResponse.json(mockUsers);
  }),

  http.get(
    'https://jsonplaceholder.typicode.com/users/:id',
    async ({ params }) => {
      await artificialDelay();
      const id = parseInt(params['id'] as string);
      const user = mockUsers.find(u => u.id === id);

      if (!user) {
        return new HttpResponse(null, { status: 404 });
      }

      return HttpResponse.json(user);
    },
  ),

  // Posts
  http.get(
    'https://jsonplaceholder.typicode.com/posts',
    async ({ request }) => {
      await artificialDelay();
      const url = new URL(request.url);
      const userId = url.searchParams.get('userId');

      if (userId) {
        const filteredPosts = mockPosts.filter(
          p => p.userId === parseInt(userId),
        );
        return HttpResponse.json(filteredPosts);
      }

      return HttpResponse.json(mockPosts);
    },
  ),

  http.get(
    'https://jsonplaceholder.typicode.com/posts/:id',
    async ({ params }) => {
      await artificialDelay();
      const id = params['id'] as string;
      const post = mockPosts.find(p => p.id === id);

      if (!post) {
        return new HttpResponse(null, { status: 404 });
      }

      return HttpResponse.json(post);
    },
  ),

  http.post(
    'https://jsonplaceholder.typicode.com/posts',
    async ({ request }) => {
      await artificialDelay();
      const newPost = (await request.json()) as Omit<Post, 'id'>;
      // Generate a unique UUID for the new post
      const post: Post = {
        ...newPost,
        id: uuidv4(),
      };

      mockPosts.push(post);
      return HttpResponse.json(post, { status: 201 });
    },
  ),

  http.put(
    'https://jsonplaceholder.typicode.com/posts/:id',
    async ({ params, request }) => {
      await artificialDelay();
      const id = params['id'] as string;
      const updatedData = (await request.json()) as Partial<Post>;
      const postIndex = mockPosts.findIndex(p => p.id === id);

      if (postIndex === -1) {
        return new HttpResponse(null, { status: 404 });
      }

      const existingPost = mockPosts[postIndex]!;

      const updatedPost: Post = {
        id: existingPost.id,
        userId: updatedData.userId ?? existingPost.userId,
        title: updatedData.title ?? existingPost.title,
        body: updatedData.body ?? existingPost.body,
      };

      mockPosts[postIndex] = updatedPost;
      return HttpResponse.json(updatedPost);
    },
  ),

  http.delete(
    'https://jsonplaceholder.typicode.com/posts/:id',
    async ({ params }) => {
      await artificialDelay();
      const id = params['id'] as string;
      const postIndex = mockPosts.findIndex(p => p.id === id);

      if (postIndex === -1) {
        return new HttpResponse(null, { status: 404 });
      }

      mockPosts.splice(postIndex, 1);
      return new HttpResponse(null, { status: 200 });
    },
  ),
];
