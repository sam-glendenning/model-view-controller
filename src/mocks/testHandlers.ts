import { http, HttpResponse } from 'msw';
import type { Post, User } from '@/shared/types';
import { mockUsers, mockPosts } from './data';
import { v4 as uuidv4 } from 'uuid';

// Create isolated copies for each test handler call to prevent mutations
const getIsolatedMockPosts = (): Post[] =>
  JSON.parse(JSON.stringify(mockPosts));
const getIsolatedMockUsers = (): User[] =>
  JSON.parse(JSON.stringify(mockUsers));

// Test handlers with no delay and NO shared state mutations
export const testHandlers = [
  // Users
  http.get('https://jsonplaceholder.typicode.com/users', async () => {
    return HttpResponse.json(getIsolatedMockUsers());
  }),

  http.get(
    'https://jsonplaceholder.typicode.com/users/:id',
    async ({ params }) => {
      const id = parseInt(params['id'] as string);
      const users = getIsolatedMockUsers();
      const user = users.find(u => u.id === id);

      if (!user) {
        return new HttpResponse(null, { status: 404 });
      }

      return HttpResponse.json(user);
    },
  ),

  // Posts
  http.get(
    'https://jsonplaceholder.typicode.com/posts',
    async ({ request }) => {
      const url = new URL(request.url);
      const userId = url.searchParams.get('userId');
      const posts = getIsolatedMockPosts();

      if (userId) {
        const filteredPosts = posts.filter(p => p.userId === parseInt(userId));
        return HttpResponse.json(filteredPosts);
      }

      return HttpResponse.json(posts);
    },
  ),

  http.get(
    'https://jsonplaceholder.typicode.com/posts/:id',
    async ({ params }) => {
      const id = params['id'] as string;
      const posts = getIsolatedMockPosts();
      const post = posts.find(p => p.id === id);

      if (!post) {
        return new HttpResponse(null, { status: 404 });
      }

      return HttpResponse.json(post);
    },
  ),

  http.post(
    'https://jsonplaceholder.typicode.com/posts',
    async ({ request }) => {
      const newPost = (await request.json()) as Omit<Post, 'id'>;
      // Generate a unique UUID for the new post
      const post: Post = {
        ...newPost,
        id: uuidv4(),
      };

      // Return the new post WITHOUT mutating global state
      return HttpResponse.json(post, { status: 201 });
    },
  ),

  http.put(
    'https://jsonplaceholder.typicode.com/posts/:id',
    async ({ params, request }) => {
      const id = params['id'] as string;
      const updatedData = (await request.json()) as Partial<Post>;
      const posts = getIsolatedMockPosts();
      const postIndex = posts.findIndex(p => p.id === id);

      if (postIndex === -1) {
        return new HttpResponse(null, { status: 404 });
      }

      // Create updated post WITHOUT mutating global state
      const updatedPost = { ...posts[postIndex], ...updatedData };
      return HttpResponse.json(updatedPost);
    },
  ),

  http.delete(
    'https://jsonplaceholder.typicode.com/posts/:id',
    async ({ params }) => {
      const id = params['id'] as string;
      const posts = getIsolatedMockPosts();
      const postIndex = posts.findIndex(p => p.id === id);

      if (postIndex === -1) {
        return new HttpResponse(null, { status: 404 });
      }

      // Return success WITHOUT mutating global state
      return new HttpResponse(null, { status: 200 });
    },
  ),
];
