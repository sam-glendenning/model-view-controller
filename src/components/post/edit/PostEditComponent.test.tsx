import { render } from '@testing-library/react';
import {
  PostEditComponent,
  type PostEditComponentProps,
} from './PostEditComponent';
import { mockPosts } from '@/mocks/data';

describe('PostEditComponent', () => {
  it('renders correctly', () => {
    const props: PostEditComponentProps = {
      formData: mockPosts[0],
      onTitleChange: jest.fn(),
      onBodyChange: jest.fn(),
      onUserIdChange: jest.fn(),
      onSave: jest.fn(),
      onCancel: jest.fn(),
      onDelete: jest.fn(),
      isSaveButtonDisabled: false,
      isUpdating: false,
      isDeleting: false,
    };

    const { asFragment } = render(<PostEditComponent {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
