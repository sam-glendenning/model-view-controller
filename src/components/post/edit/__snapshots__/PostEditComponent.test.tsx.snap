// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostEditComponent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-okv3ew-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1j5gcdd-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-1"
            id="mui-1-label"
          >
            Title
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              type="text"
              value="Test Post Title"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-r01yff-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-sl37lx-MuiNotchedOutlined-root"
              >
                <span>
                  Title
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1j5gcdd-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-2"
            id="mui-2-label"
          >
            Content
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-2"
              rows="4"
              style="height: 0px; overflow: hidden;"
            >
              This is a test post body with some content to display.
            </textarea>
            <textarea
              aria-hidden="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-r01yff-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-sl37lx-MuiNotchedOutlined-root"
              >
                <span>
                  Content
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-3"
            id="mui-3-label"
          >
            User ID
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-3"
              type="number"
              value="1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-r01yff-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-sl37lx-MuiNotchedOutlined-root"
              >
                <span>
                  User ID
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1j9t6id-MuiCardActions-root"
    >
      <div
        class="MuiBox-root css-1i27l4i"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-11qr2p8-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SaveIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3m3-10H5V5h10z"
              />
            </svg>
          </span>
          Save
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CancelIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2m5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12z"
              />
            </svg>
          </span>
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <button
        aria-label="delete post"
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError css-1ir7th6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
            />
          </svg>
        </span>
        Delete
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
